utils/bot/poster/default_poster.go

```diff
@@ -0,0 +1,99 @@
+package poster
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+type defaultPoster struct {
+	pluginAPI plugin.API
+	id        string
+}
+
+func NewPoster(api plugin.API, id string) Poster {
+	return &defaultPoster{
+		pluginAPI: api,
+		id:        id,
+	}
+}
+
+// DM posts a simple Direct Message to the specified user
+func (p *defaultPoster) DM(mattermostUserID, format string, args ...interface{}) (string, error) {
+	postID, err := p.dm(mattermostUserID, &model.Post{
+		Message: fmt.Sprintf(format, args...),
+	})
+	if err != nil {
+		return "", err
+	}
+	return postID, nil
+}
+
+// DMWithAttachments posts a Direct Message that contains Slack attachments.
+// Often used to include post actions.
+func (p *defaultPoster) DMWithAttachments(mattermostUserID string, attachments ...*model.SlackAttachment) (string, error) {
+	post := model.Post{}
+	model.ParseSlackAttachment(&post, attachments)
+	return p.dm(mattermostUserID, &post)
+}
+
+func (p *defaultPoster) dm(mattermostUserID string, post *model.Post) (string, error) {
+	channel, err := p.pluginAPI.GetDirectChannel(mattermostUserID, p.id)
+	if err != nil {
+		p.pluginAPI.LogInfo("Couldn't get bot's DM channel", "user_id", mattermostUserID)
+		return "", err
+	}
+	post.ChannelId = channel.Id
+	post.UserId = p.id
+	sentPost, err := p.pluginAPI.CreatePost(post)
+	if err != nil {
+		return "", err
+	}
+	return sentPost.Id, nil
+}
+
+// Ephemeral sends an ephemeral message to a user
+func (p *defaultPoster) Ephemeral(userId, channelId, format string, args ...interface{}) {
+	post := &model.Post{
+		UserId:    p.id,
+		ChannelId: channelId,
+		Message:   fmt.Sprintf(format, args...),
+	}
+	_ = p.pluginAPI.SendEphemeralPost(userId, post)
+}
+
+func (p *defaultPoster) DMUpdate(postID, format string, args ...interface{}) error {
+	post, appErr := p.pluginAPI.GetPost(postID)
+	if appErr != nil {
+		return appErr
+	}
+
+	post.Message = fmt.Sprintf(format, args...)
+	_, appErr = p.pluginAPI.UpdatePost(post)
+	if appErr != nil {
+		return appErr
+	}
+
+	return nil
+}
+
+func (p *defaultPoster) DeletePost(postID string) error {
+	appErr := p.pluginAPI.DeletePost(postID)
+	if appErr != nil {
+		return appErr
+	}
+	return nil
+}
+
+func (p *defaultPoster) UpdatePost(post *model.Post) error {
+	_, appErr := p.pluginAPI.UpdatePost(post)
+	if appErr != nil {
+		return appErr
+	}
+	return nil
+}
+
+func (p *defaultPoster) UpdatePosterID(id string) {
```

Comment:
This came from a need on how the MSCalendar plugin is built. You may want to set up everything before ensuring the bot, and you do not have the user ID until you have the bot. Nevertheless, this might be solved restructuring the code in MSCalendar.

We could tie a poster with a bot, but I thought that it was more general and interesting to have the poster be a poster from any userID you want to use. This way you can post things as a user instead of posting as a bot.